### PR TITLE
Add GitHub auth

### PR DESCRIPTION
**Pull request by xobb1t**
For future use we adding user authentication with github.
At this point we need to fix layout (link in header is not stylized yet) and add gravatar.
Also I should to create tests for authentication process.

### CODE REVIEW

**Comment by xobb1t** [comment]
okay, i added gravatar. Hope that @nyumatova will help me with layout. @yumike, please review python code.

File: project/oauth/middleware.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from django.utils.functional import SimpleLazyObject
+
```
Comment (yumike) [inline project/oauth/middleware.py:3]: Why this blank line is necessary?

File: project/oauth/middleware.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+from .models import User
+
+
+class AnonymousUser(object):
```
Comment (yumike) [inline project/oauth/middleware.py]: I think that this model class must be in models.py

Comment (xobb1t) [inline project/oauth/middleware.py]: Yeah, I will move it.

File: project/oauth/middleware.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        return False
+
+    def is_anonymous(self):
+        return True
+
+
+class UserMiddleware(object):
+
+    def process_request(self, request):
+        def get_user():
```
Comment (yumike) [inline project/oauth/middleware.py]: You can move this function out of `process_request` method (to the module level), and use lambda for `SimpleLazyObject`. `process_request` method will be probably simpler in this case:

```
def process_request(self, request):
    request.user = SimpleLazyObject(lambda: get_user(request))
```

File: project/oauth/models.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from django.db import models
+from github import Github
+
+
+class User(models.Model):
+
+    identificator = models.CharField(max_length=255, unique=True)
```
Comment (yumike) [inline project/oauth/models.py]: `identificator` => `identifier`. Also: how this is different from `id`? I mean, `User` has `id` field and `identificator` field - names are the same, this is confusing.

Comment (xobb1t) [inline project/oauth/models.py]: Identifier is used to store Github User ID. Is it ok to use it as primary key ?

File: project/oauth/models.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+from django.db import models
+from github import Github
+
+
+class User(models.Model):
```
Comment (yumike) [inline project/oauth/models.py:14]: All fields has `255` length. Is this justified?

File: project/oauth/urls.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+from django.conf.urls.defaults import url, patterns
+
+
+urlpatterns = patterns('oauth.views',
```
Comment (yumike) [inline project/oauth/urls.py]: Do not use implicit imports here, please. Just import view functions at the top of the module.

File: project/oauth/models.py (hunk 5 above)
Comment (xobb1t) [inline project/oauth/models.py:14]: My point is not to depent on provider data length restrictions.

File: project/oauth/urls.py (hunk 6 above)
Comment (xobb1t) [inline project/oauth/urls.py]: ok

File: project/oauth/middleware.py (hunk 1 above)
Comment (xobb1t) [inline project/oauth/middleware.py:3]: Removed

File: project/oauth/middleware.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+from django.utils.functional import SimpleLazyObject
+from .models import User, AnonymousUser
+
+
+def get_user(request):
+    user_id = request.session.get('user_id')
+    if not user_id:
```
Comment (yumike) [inline project/oauth/middleware.py]: If you expect `None` or something other, please check for `None` or `not None`.

File: project/oauth/models.py (hunk 4 above)
Comment (yumike) [inline project/oauth/models.py]: Let's rename it to `github_id` and set `unique` to `True`.

Comment (xobb1t) [inline project/oauth/models.py]: Ok, done

**Comment by xobb1t** [comment]
Due to plans about moving to flask in near future, I think we don't need to change layout right now. So, I'm done with it, I think.

**Comment by yumike** [comment]
Checked it, and it works! :beer: :beer: :beer:
Let's merge it.

**Pull request merged**